doc/developer/design/20220330_persist.md

```diff
@@ -0,0 +1,1139 @@
+they don’t get a success message back. A restarted writer **must still use a new
+writer ID**, to ensure that we don’t write duplicates.
+
+The part that reads from the WAL and writes to persistence can be thought of as
+a source, where the WAL is the source system (see below) and we don’t need to do
+timestamping.
+
+### Sources
+
+A source needs to roughly do this:
+- Read from an external system
+- Attach a timestamp to raw data
+- Write the timestamped data to persistence (a shard)
+
+The difficulty here is ensuring that a given source instance writes the same (in
+terms of definiteness) updates as all the other instances. Both when it comes to
+instances across time, after restarts, and with concurrently running instances.
+To solve this, we need two things: source messages must come with some kind of
+identifier that we can use to look up a timestamp that we attach (think Kafka
+partition-offset pairs), let’s call this _source timestamp_ and we need a
+definite collection of timestamp bindings where we look up the assigned
+timestamp for a given source timestamp. We will sketch two different approaches
+for this. There might be more.
+
+For both approaches, the collection of timestamp bindings works like this: the
+schema of bindings is `(<source timestamp>, <ts>)` and a binding `(s_ts, ts)`
+tells us that any update with a source timestamp that is less-equal `s_ts` gets
+assigned the timestamp `ts`, unless there is another binding (with a lower
+source timestamp) that already covers a given source timestamp. So the
+collection of timestamp bindings define ranges of source timestamps that get
+assigned the same timestamp.
+
+Both approaches have in common that there needs to be durable consensus about
+what the timestamp bindings for given source updates should be. The two sketched
+approaches differ in how those bindings are minted, though. We will call the
+component that reads from an external source, applies bindings, and writes the
+timestamped data to a persist shard a _source writer_.
+
+#### Centralized Timestamp Minter
+
+**This approach assumes that it is possible to figure out the "high watermarks"
+of available data, in terms of their source timestamp.** The centralized minter
+can then mint bindings of the current "real-time clock" and write them down.
+This writing down needs some kind of consensus/exactly once to ensure that there
+is a consistent collection of bindings that all the source writers use.
+
+All the source writers need to do with this model is observe the minted
+bindings, read updates from the source, attach bindings to updates and write
+them to persistence.
+
+#### Distributed Minting (but still with consensus, so also centralized in a way) (title of this section TBD… 😅)
+
+With this approach, each source writer can read data from the source and
+whenever it encounters a source timestamp for which it doesn’t know the binding,
+it will "ask" a centralized component for a binding. That component can be
```

Comment:
yeah I think that going through a central component like this makes sense -- and I think having the timestamps minted on request from the workers is preferable.  The amount of work done by the centralized piece corresponds to the activity of the sources rather than just how many sources (since it'd have to poll every `$interval` to see if there are updates).  I'm not able to think of a reason we'd need a big temporal gap in data to have many timestamps assigned to it

Reply:
👍